File: 0006-zigzag-conversion/0006-zigzag-conversion.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function(s, numRows) {
    if (numRows < 2 || s.length < numRows) return s;
    var rows = new Array(numRows).fill('');
    var forward = true;
    var idx = 0;
    for (var i = 0; i < s.length; i++) {
        rows[idx] += s[i]
        forward ? idx++ : idx--
        if (idx === numRows - 1 || idx === 0) forward = !forward
    }
    return rows.join('');
};

File: 0006-zigzag-conversion/0006-zigzag-conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1 or numRows >= len(s):
            return s
        
        l = [''] * numRows
        idx = 0
        inc = 1
        
        for let in s:
            l[idx] += let
            if idx == numRows - 1:
                inc = -1
            elif idx == 0:
                inc = 1
            idx += inc
        return ''.join(l)